# SQL 4 Data Science: Predict Project
## Bhejane Online Trading Store

### Student Starter Notebook

© Explore Data Science Academy

### Honour Code

I {**YOUR NAME**, **YOUR SURNAME**}, confirm - by submitting this document - that the solutions in this notebook are a result of my own work and that I abide by the EDSA honour code (https://drive.google.com/file/d/1QDCjGZJ8-FmJE3bZdIQNwnJyQKPhHZBn/view?usp=sharing).

Non-compliance with the honour code constitutes a material breach of contract.

<a id='Context'></a>
### 1. Context

The Bhejane trading store is an online retailer specialising in Covid essential items. The store has recently been struggling with the management of its database-related inventory system. Luckily for them, you've been hired as a consultant to fix the problem. 

Its time to flex your Ninja SQL skills! 
 
**Your mission, should you choose to accept it:**

You will receive a denormalised database consisting of two tables. Your task is to transform the database such that is in third normal form (3NF). To ensure that a consistent normalisation process is followed by each student, we have attached an Entity Relationship Diagram (ERD) which is a guideline on what tables need to be produced. Once you have normalised the database, you will be required to complete an MCQ test found under the 'Predict' tab of Athena.  

Though the normalisation process, you will be guided into acknowledging client (Bhenjane Trading) use-cases for the database, and dealing with data anomalies in SQL. 

![Bhejane company logo](https://raw.githubusercontent.com/Explore-AI/Pictures/master/SQL4DS/SQL_Predict/Bhejane.png)

<div align="center" >
    Bhejane, or the Black Rhino. Image by <a href="https://commons.wikimedia.org/wiki/File:Black_Rhino_(Diceros_bicornis)_browsing_..._(46584052962).jpg">Wikimedia Commons</a>
</div>

<a id='Imports'></a>
### 2. Imports
Please use the below command to install sql_magic, this is the package that will assist you with SQL syntax hightlighting.
* pip install sql_magic

Remember to start each new cell with " %%read_sql "

In [13]:
# DO NOT EDIT THIS CELL
import sqlite3
import csv
from sqlalchemy import create_engine
%load_ext sql_magic

# Load SQLite database
engine  = create_engine("sqlite:///bhejane.db")
%config SQL.conn_name ='engine'

The sql_magic extension is already loaded. To reload it, use:
  %reload_ext sql_magic


<a id='Data_description'></a>
### 3. Data description

The original database consist of 2 tables. 
* Product Table
* Transaction Table

The `Product` table consists of the stock of all inventory that Bhejane has on hand currently, or has had on hand historically. Items which are in this table are able to be purchased, and a record of all sales (transactions) in 2020 is notorised in the `Transactions` table. To link the tables - the `barcode` can be used. Any item in the `Transactions` table, must therefore appear in the `Products` table. 

In [14]:
# DO NOT EDIT THIS CELL
import pandas as pd
data_description = pd.read_excel('Data Description.xlsx')
data_description

Table Name         Column Name  \
0       Products               Width   
1       Products              Length   
2       Products              Height   
3       Products             Barcode   
4       Products            Quantity   
5       Products               Brand   
6       Products      NavigationPath   
7       Products              Colour   
8       Products        StockCountry   
9       Products  ProductDescription   
10      Products            PackType   
11      Products        Volume_litre   
12      Products            Warranty   
13      Products           Weight_kg   
14      Products     ItemDescription   
15      Products               Price   
16  Transactions              CartID   
17  Transactions             Barcode   
18  Transactions               Total   
19  Transactions            UserName   
20  Transactions         InvoiceDate   

                                          Desciption   
0                 Width of the product once assembled  
1                Length of the product once assembled  
2                Height of the product once assembled  
3                      The unique product identifier   
4                            Number of goods in stock  
5      Product brand name relating to product company  
6                 Navigation path to specific product  
7                 Name default colour for the product  
8             Country where the stock was bought from  
9                            Descriptive product name  
10           The way in which the product is packaged  
11              The quantity of liquid based products  
12                              Product warranty type  
13                              Weight of the product  
14            Description of contents of the package   
15                   The cost of the product in Rands  
16  Identifies an individual basket from a user pu...  
17               Barcode for item that user purchased  
18                      Total value of a transaction   
19                  Name of customer purchasing stock  
20               Date on which the product was bought

<a id='Setting_up'></a>
### 4. Setting up the database

In [15]:
#DO NOT EDIT THIS CELL
conn = sqlite3.connect('bhejane.db')
cursor = conn.cursor()

#### Creating both Products and Transaction Tables

In [16]:
%%read_sql
--DO NOT EDIT THIS CELL
DROP TABLE IF EXISTS "Products";
DROP TABLE IF EXISTS "Transactions";

CREATE TABLE "Products" (
    "Width"   REAL,
    "Length"  REAL,
    "Height"  REAL,
    "Barcode" VARCHAR(150),
    "Quantity" REAL,
    "Brand" VARCHAR(150), 
    "NavigationPath" VARCHAR(150),
    "Colour" VARCHAR(150),
    "StockCountry" VARCHAR(150),
    "ProductDescription" VARCHAR(150),
    "PackType" VARCHAR(150), 
    "Volume_litre" REAL, 
    "Warranty" VARCHAR(150), 
    "Weight_kg" REAL,
    "ItemDescription" VARCHAR(150), 
    "Price" REAL
);


CREATE TABLE "Transactions" (
    "CartID" INTEGER,
    "Barcode" VARCHAR(150), 
    "Total" REAL,
    "UserName" VARCHAR(150), 
    "InvoiceDate" DATETIME
);

Query started at 10:48:06 PM South Africa Standard Time; Query executed in 0.01 m

#### Load Data into Product and Transaction tables

Although we have not yet covered python in the sprint at EXPLORE - you have additional resources at your disposal, one of which happens to be a data engineering expert, and she has created a brief script for you which can be used to extract the data from the csv files, and load it into a sqlite database. 

This will be included in your submission, so that the end-to-end processing of the data is visible, and repeatable for any additional consultants who may be brought on board at a later stage. 


In [17]:
# DO NOT EDIT CELL
#Load data into Product table
with open('bhejane_covid_essentials_Products.csv','r') as fin: # `with` statement available in 2.5+
    # csv.DictReader uses first line in file for column headings by default
    dr = csv.DictReader(fin) # comma is default delimiter
    to_db = [(i['Width'],i['Length'],i['Height'], i['Barcode'], i['Quantity'], i['Brand'], i['NavigationPath'], i['Colour'], i['StockCountry'], i['ProductDescription'],i['PackType'],i['Volume_litre'],i['Warranty'],i['Weight_kg'],i["ItemDescription"],i['Price']) for i in dr]

cursor.executemany("INSERT INTO Products VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?);", to_db)
conn.commit()

with open('bhejane_covid_essentials_Transactions.csv','r') as fin: # `with` statement available in 2.5+
    # csv.DictReader uses first line in file for column headings by default
    dr = csv.DictReader(fin) # comma is default delimiter
    to_db = [(i['CartID'],i['Barcode'], i['Total'], i['UserName'], i['InvoiceDate']) for i in dr]

cursor.executemany("INSERT INTO Transactions VALUES (?, ?, ?, ?, ?);", to_db)
conn.commit()

<a id='ERD'></a>
### 5. Denormalized Database Tables

<br>

<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/SQL4DS/SQL_Predict/Denormalized_Tables.png" alt="Denormalized Tables" border="0">

#### 5.1 Explore the Denormalized Tables
Familiarize yourself with the data given in the two tables by writing your own SQL queries to explore properties of the dataset.i.e Look for data inconsistencies, anormalies, redundancies etc to guide your normalization process. 

This section will not be graded - think of it as your own scratch pad.



In [18]:
%%read_sql
-- #Use this section to explore the database by writing your own queries
SELECT * FROM Products
LIMIT 5;

Query started at 10:48:14 PM South Africa Standard Time; Query executed in 0.02 m

Width Length Height      Barcode  Quantity         Brand  \
0                        300507946     493.0     Hikvision   
1                        300507946     493.0     Hikvision   
2                      10325354918     467.0           ZEE   
3                      27131187035     275.0  Estee Lauder   
4                      27131187035     275.0  Estee Lauder   

                                      NavigationPath  Colour  StockCountry  \
0  Computers & Tablets / Smart Home & Connected L...                         
1  Computers & Tablets / Smart Home & Connected L...                         
2                    Fashion / Accessories / Scarves    Grey                 
3  Beauty / Luxury Beauty / Makeup / Face / Found...  Fresco  South Africa   
4  Beauty / Luxury Beauty / Shop By Brand / Estee...  Fresco  South Africa   

                              ProductDescription PackType  Volume_litre  \
0  Hikvision 1080P 2MP Turbo HD IR Bullet Camera                    0.0   
1  Hikvision 1080P 2MP Turbo HD IR Bullet Camera                    0.0   
2                       ZEE 3-in-1 Unisex Gaiter                    0.0   
3  Estee Lauder Double Wear Stay In Place Makeup                    0.0   
4  Estee Lauder Double Wear Stay In Place Makeup                    0.0   

             Warranty Weight_kg                    ItemDescription  Price  
0  Limited (6 months)            1 x Hikvision 1080P Bullet camera  399.0  
1  Limited (6 months)                                       Manual  399.0  
2  Limited (6 months)                                               139.0  
3      Non-Returnable                                               655.0  
4      Non-Returnable                                               655.0

### 6. Normalize the given Database tables to the 1st Normal Form (1NF)

Given the below below target ERD create new tables such the the database conforms to the 1st Normal Form

![1st Normal Form](https://raw.githubusercontent.com/Explore-AI/Pictures/master/SQL4DS/SQL_Predict/1stNF.png)


It is suggested that you create all your tables before attempting to populate them with data, this will help reduce errors that might creep up due to logical dependencies.

#### 6.1 Create the tables required for the 1st Normal Form
Use the above ERD sketch to create the required tables. Please label your tables as they appear in the ERD sketch

In [19]:
%%read_sql
--#Create tables required for 1NF
DROP TABLE IF EXISTS "Products_1NF";
DROP TABLE IF EXISTS "Transactions_1NF";

CREATE TABLE "Products_1NF"(
    "Barcode" VARCHAR(150),
    "NavigationPath" VARCHAR(150),
    "ItemDescription" VARCHAR(150),
    "Colour" VARCHAR(150),
    "ProductDescription" VARCHAR(150) NOT NULL,
    "Brand" VARCHAR(150),
    "Price" REAL NOT NULL,
    "Quantity" INTEGER NOT NULL,
    "PackType" VARCHAR(150),
    "Warranty" VARCHAR(150),
    "StockCountry" VARCHAR(150),
    "Weight_kg" REAL,
    "Volume_litre" REAL,
    "Length" REAL,
    "Width" REAL,
    "Height" REAL,
    PRIMARY KEY("Barcode","NavigationPath","ItemDescription")
);

CREATE TABLE "Transactions_1NF" (
    "CartID" INTEGER NOT NULL,
    "Barcode" VARCHAR(150) NOT NULL,
    "UserName" VARCHAR(150) NOT NULL,
    "InvoiceDate" DATETIME NOT NULL,
    "Total" REAL NOT NULL,
    FOREIGN KEY ("Barcode") REFERENCES "Products_1NF" ("Barcode"),
    PRIMARY KEY("CartID", "Barcode", "UserName")
);

Query started at 10:48:21 PM South Africa Standard Time; Query executed in 0.02 m

#### 6.2 Populate the tables you have create in the above section.

Populate the tables such that the database conforms to the 1st Normal Form

In [20]:
%%read_sql
--#Populate the 1NF tables
DELETE FROM "Products_1NF";
DELETE FROM "Transactions_1NF";

INSERT INTO "Products_1NF" ("Barcode","NavigationPath","ItemDescription","Colour","ProductDescription","Brand",
                            "Price","Quantity","PackType","Warranty","StockCountry","Weight_kg",
                            "Volume_litre","Length","Width","Height")
SELECT DISTINCT
     Barcode
    ,NavigationPath
    ,ItemDescription
    ,Colour
    ,ProductDescription
    ,Brand
    ,Price
    ,Quantity
    ,PackType
    ,Warranty
    ,StockCountry
    ,Weight_kg
    ,Volume_litre
    ,Length
    ,Width
    ,Height 
FROM 
    Products;

INSERT INTO "Transactions_1NF"("CartID","Barcode","UserName","InvoiceDate","Total")
SELECT DISTINCT 
    CartID
    ,Barcode
    ,UserName
    ,InvoiceDate
    ,Total
FROM
    Transactions;
    

Query started at 10:48:26 PM South Africa Standard Time; Query executed in 0.01 m

###  7. Convert the database into its 2nd Normal Form (2NF).

Please label your tables as they appear in the ERD sketch.

#### 7.1 Entity Relationship Diagram

![2nd Normal Form](https://raw.githubusercontent.com/Explore-AI/Pictures/master/SQL4DS/SQL_Predict/2ndNF.png)

#### 7.2 2NF Requirements
To transition from 1NF to 2NF, we need all columns in all tables to not have a partial dependancy on the PK of the table. This means that any tables which have a composite key e.g. `PRIMARY KEY("Barcode","NavigationPath","ItemDescription")` cannot have columns in the table which are dependant on only `Barcode`, `NavigationPath`, or `ItemDescription`. 

In moving to 2NF we will seperate out each of these as a primary key (for their own respective tables), and any columns dependant on one of these columns but not the others, will go into that table. The only columns remaining in the tables will be fully functionally dependant on the primary key of the table. For example, 

Thereby, we remove the partial dependancy, and will be in 2NF. We are instructed to create a unique identifier for the `Transactions_2NF`. 

**Second normal form notes**

<br>$\bullet$ Notice that `Total` is only dependant on `CartID` and not on `<CartID,barcode>`. This is a partial dependancy
<br>$\bullet$ Second normal form required no *partial functional dependancy* on the key. 
<br>$\bullet$ This can be achieved in 2 ways: create a unique identifier per row (so that there is a simple primary key), or break the table up into only the columns which are fully dependant on the composite key. 
<br>$\bullet$ We are going to opt for the easier solution, and create a unique simple primary key on the `Transactions_1NF`. We can create a unique primary key by concatenating `CartID||barcode`
<br>$\bullet$ NB - this approach does make it more challenging to get into 3NF! Which would be a good argument to opt for the latter approach. 

#### 7.3 Data Anomalies

You may stumble into an error when you insert the data which violates this constraint. You query this with the data owner, and you are told that there are three transactions in the database which were errors. You are informed that *all transactions in the database correspond to products are in the database*.

These entries were captured incorrectly, and must be removed. How will you deal with this? 

In [21]:
%%read_sql
select count(distinct barcode) from Transactions_1NF where barcode not in (select barcode from products_1NF)

Query started at 10:48:32 PM South Africa Standard Time; Query executed in 0.00 m

count(distinct barcode)
0                        2

### <font color='turquoise'>Action: remove the affected entries with a `delete` query </font>

In [22]:
%%read_sql
delete from Transactions_1NF
WHERE barcode not in (SELECT barcode FROM products_1NF)

Query started at 10:49:03 PM South Africa Standard Time; Query executed in 0.00 m

In [23]:
%%read_sql
-- Should show no entries!
select count(distinct barcode) from Transactions_1NF where barcode not in (select barcode from products_1NF)

Query started at 10:49:06 PM South Africa Standard Time; Query executed in 0.00 m

count(distinct barcode)
0                        0

#### 7.4 FK Requirements

_Note: sqlite does not always enforce foreign key constraints (and it is even more inconsistent with magic commands to enforce FK constraints!), so it is critical that you understand the purpose of the foreign key constraint and therefore are able to enforce the FK constraint yourself!_

To enforce data integrity, and to ensure that strange things cannot happen - we will have a FK constraint. A foreign key forces all values of the FK, to be limited to only the values which exist as entries in the PK column of the table for which the aforementioned table is a foreign key to. For example, we cannot sell a product (in our Transactions table) which we do not have in our Products table. 

You will need to work around this by removing the erroneous data from the tables before continuing, since this would violate the FK constraint between the Transactions_2NF and Products_2NF table otherwise. 

_Hint: Identify the observations in the Transactions_1NF table, and remove them from there._

#### 7.5 Client Use-Case Requirements

The data owner has advised that there were several instances of incorrect data capturing. Whenever data was partially captured, it was redone, but the partially-complete entries have not been removed from the database. 

For all tables which are going to be used, consider the table utility relative to the Products (Products_2NF) table. You are tasked with removing 'useless' entries as you take the data from 1NF to 2NF. For the sake of ease-of-use for the client, we will removing all rows in the 2NF tables `PackageContents_2NF` and `Colours_2NF` which have an empty value for the attributes that the client will use them for - these are the `ItemDescription` and `Colour` columns, respectively. 

_Hint: Make inserts into the other 2NF (`PackageContents_2NF` and `Colours_2NF`) tables using a `where` clause to remove instances of the value being an empty string, or a `null` value. This will need to be considered for the `ItemDescription` and for the `Colour` when data is inserted into the corresponding tables above (i.e. only insert data into the `PackageContents_2NF` which has a value for the `ItemDescription`, and only insert data into the `Colours_2NF` table which has a value for `Colour`_

### <font color='turquoise'>Action: Investigate these anomalies for the tables mentioned </font>

In [24]:
%%read_sql 

SELECT DISTINCT 
    ItemDescription,
    PackType,
    Warranty 
FROM
    Products
WHERE ItemDescription = '';

Query started at 10:49:13 PM South Africa Standard Time; Query executed in 0.00 m

ItemDescription PackType              Warranty
0                              Limited (6 months)
1                                  Non-Returnable
2                             Limited (12 months)
3                            Limited (120 months)
4                            Limited (180 months)
5                             Limited (18 months)
6                            Supplier (12 months)
7                    Single    Limited (6 months)
8                             Limited (24 months)
9                                                
10                            Limited (36 months)
11                                Full (6 months)

### Let's do it! Construct the database above in second normal form

In [25]:
%%read_sql
--#Create tables required for 2NF

DROP TABLE IF EXISTS "Products_2NF";
DROP TABLE IF EXISTS "Transactions_2NF";
DROP TABLE IF EXISTS "Navigation_2NF";
DROP TABLE IF EXISTS "PackageContents_2NF";
DROP TABLE IF EXISTS "Colours_2NF";

CREATE TABLE Products_2NF(
    "RegistryID" INTEGER NOT NULL,
    "Barcode" VARCHAR(150) NOT NULL,
    "ProductDescription" VARCHAR(150) NOT NULL,
    "Brand" VARCHAR(150),
    "Price" REAL NOT NULL,
    "Quantity" INTEGER NOT NULL,
    "StockCountry" VARCHAR(150),
    "Weight_kg" REAL,
    "Volume_litre" REAL,
    "Length" REAL,
    "Width" REAL,
    "Height" REAL,
    "PathID" INTEGER,
    "ItemID" INTEGER,
    "ColourID" INTEGER,
    PRIMARY KEY(RegistryID AUTOINCREMENT),
    CONSTRAINT fk_nav FOREIGN KEY('PathID') REFERENCES Navigation_2NF ('PathID'),
    CONSTRAINT fk_pack FOREIGN KEY('ItemID') REFERENCES PackageContents_2NF ('ItemID'),
    CONSTRAINT fk_col FOREIGN KEY('ColourID') REFERENCES Colours_2NF ('ColourID') 
);

CREATE TABLE "Transactions_2NF" (
    "CartID||Barcode||UserName" VARCHAR(150) NOT NULL,
    "CartID" INTEGER NOT NULL,
    "Barcode" VARCHAR(150) NOT NULL,
    "UserName" VARCHAR(150) NOT NULL,
    "InvoiceDate" DATETIME NOT NULL,
    "Total" REAL NOT NULL,
    PRIMARY KEY("CartID||Barcode||UserName")
    CONSTRAINT fk_trans FOREIGN KEY('Barcode') REFERENCES Products_2NF ('Barcode')
);


CREATE TABLE "Navigation_2NF"(
    "PathID" INTEGER NOT NULL,
    "NavigationPath" VARCHAR(150),
    PRIMARY KEY("PathID" AUTOINCREMENT)
);

CREATE TABLE "Colours_2NF"(
    "ColourID" INTEGER NOT NULL,
    "Colour" VARCHAR(150),
    PRIMARY KEY("ColourID" AUTOINCREMENT)
);

CREATE TABLE "PackageContents_2NF"(
    "ItemID" INTEGER NOT NULL,
    "ItemDescription" VARCHAR(150),
    "PackType" VARCHAR(150),
    "Warranty" VARCHAR(150),
    PRIMARY KEY("ItemID" AUTOINCREMENT)
);


Query started at 10:49:20 PM South Africa Standard Time; Query executed in 0.00 m

### Let's do it! _Populate_ the database above in second normal form

Firstly, let's consider the entries which are `null` and `= ''`. And we will only insert the relevant entries. 

In [26]:
%%read_sql
-- #Populate the tables so that they conform to 2NF

DELETE FROM "Products_2NF";
DELETE FROM "Transactions_2NF";
DELETE FROM "Navigation_2NF";
DELETE FROM "PackageContents_2NF";
DELETE FROM "Colours_2NF";


INSERT INTO "Navigation_2NF"("NavigationPath")
SELECT DISTINCT 
    NavigationPath 
FROM
    Products_1NF;
    
INSERT INTO "PackageContents_2NF"("ItemDescription","PackType","Warranty")
SELECT DISTINCT 
    ItemDescription,
    PackType,
    Warranty 
FROM
    Products_1NF
WHERE ItemDescription <> ''
        ;
    
INSERT INTO "Colours_2NF"("Colour")
SELECT DISTINCT 
    Colour
FROM
    Products_1NF
WHERE Colour <> ''
        ;

Query started at 10:49:24 PM South Africa Standard Time; Query executed in 0.01 m

A more complicated scenario below as we insert into `Products_2NF`. We will insert from the Products_1NF table. We can use LEFT JOIN, because the data in each of these tables that `Products_1NF` has a FK to, originally came from `Products_1NF`, so we are gauranteed to get matches back. Normally however, what you would get from the FK restriction is the requirement to use _parent-table_ `LEFT JOIN` _child-table_. 

In [27]:
%%read_sql 

INSERT INTO "Products_2NF" ("PathID","ItemID","ColourID","Barcode","ProductDescription","Brand","Price",
                            "Quantity","StockCountry","Weight_kg","Volume_litre",
                            "Length","Width","Height")
SELECT
    PathID,
    ItemID,
    ColourID,
    Products_1NF.Barcode,
    ProductDescription,
    Brand,
    Price,
    Quantity,
    StockCountry,
    Weight_kg, 
    Volume_litre, 
    Length,
    Width,
    Height 
FROM 
    Products_1NF
LEFT JOIN Navigation_2NF ON Products_1NF.NavigationPath = Navigation_2NF.NavigationPath 
LEFT JOIN PackageContents_2NF ON Products_1NF.ItemDescription = PackageContents_2NF.ItemDescription
LEFT JOIN Colours_2NF ON Products_1NF.Colour = Colours_2NF.Colour

Query started at 10:49:32 PM South Africa Standard Time; Query executed in 0.01 m

Last but not least, we insert into the `Transactions_2NF`table. Recall that we created a new PK in this table, and pay attention to how we are inserting values into the table. 

In [28]:
%%read_sql 

INSERT INTO "Transactions_2NF"("CartID||Barcode||UserName","CartID","Barcode","UserName","InvoiceDate","Total")
SELECT DISTINCT 
    CartID||Transactions_1NF.Barcode||UserName
    ,CartID
    ,Transactions_1NF.Barcode
    ,UserName
    ,InvoiceDate
    ,Total
FROM
    Transactions_1NF
LEFT JOIN Products_2NF ON Transactions_1NF.Barcode = Products_2NF.Barcode ;


Query started at 10:49:40 PM South Africa Standard Time; Query executed in 0.00 m

<a id='Target_ERD'></a>
### 8. Convert the table into its 3rd Normal Form (3NF)

Please label your tables as they appear in the ERD sketch. In the same fashion as done in 2NF, note which tables need to have data inserted `where column <> '' and column is not null`.

<br>

![3rd Normal Form](https://raw.githubusercontent.com/Explore-AI/Pictures/master/SQL4DS/SQL_Predict/3rdNF.png)

In [29]:
%%read_sql

--#Create tables required for 2NF
DROP TABLE IF EXISTS "Transactions_3NF";
DROP TABLE IF EXISTS "Carts_3NF";
DROP TABLE IF EXISTS "Products_3NF";
DROP TABLE IF EXISTS "Users_3NF";
DROP TABLE IF EXISTS "Navigation_3NF";
DROP TABLE IF EXISTS "PackageContents_3NF";
DROP TABLE IF EXISTS "Colours_3NF";
DROP TABLE IF EXISTS "Brands_3NF";
DROP TABLE IF EXISTS "Locations_3NF";

CREATE TABLE "Products_3NF" (
    "RegistryID" INTEGER NOT NULL,
    "Barcode" VARCHAR(150) NOT NULL,
    "ProductDescription" VARCHAR(150) NOT NULL,
    "Price" REAL NOT NULL,
    "Quantity" INTEGER NOT NULL,
    "Weight_kg" REAL,
    "Volume_litre" REAL,
    "Length" REAL,
    "Width" REAL,
    "Height" REAL,
    "PathID" INTEGER,
    "ItemID" INTEGER,
    "ColourID" INTEGER,
    "BrandID" INTEGER,
    "LocationID" INTEGER,
    PRIMARY KEY("RegistryID" AUTOINCREMENT),
    FOREIGN KEY("PathID") REFERENCES "Navigation_3NF" ("PathID"),
    FOREIGN KEY("ItemID") REFERENCES "PackageContents_3NF" ("ItemID"),
    FOREIGN KEY("ColourID") REFERENCES "Colours_3NF" ("ColourID") ,
    FOREIGN KEY ("BrandID") REFERENCES "Brands_3NF" ("BrandID"),
    FOREIGN KEY ("LocationID") REFERENCES "Locations_3NF" ("LocationID")
);


CREATE TABLE "Carts_3NF" (
    "CartID" INTEGER NOT NULL,
    "InvoiceDate" DATETIME NOT NULL,
    "Total" REAL NOT NULL,
    PRIMARY KEY ("CartID")
);


CREATE TABLE "Users_3NF"(
    "UserID" INTEGER NOT NULL,
    UserName VARCHAR(50) NOT NULL,
    PRIMARY KEY("UserID" AUTOINCREMENT)
);


Query started at 10:49:48 PM South Africa Standard Time; Query executed in 0.01 m

### <font color='turquoise'>Action: proceed with the construction of the remaining tables mentioned </font>

In [30]:
%%read_sql
CREATE TABLE "Transactions_3NF"(
    "CartID||Barcode||UserName"  VARCHAR(150),
    "CartID" INTEGER NOT NULL,
    "Barcode" VARCHAR(150) NOT NULL,
    "UserID"  INTEGER NOT NULL,
    PRIMARY KEY ("CartID||Barcode||UserName"),
    FOREIGN KEY ("Barcode") REFERENCES "Products_3NF"("UserID"),
    FOREIGN KEY ("CartID") REFERENCES "Carts_3NF" ("CartID"),
    FOREIGN KEY ("UserID") REFERENCES "Users_3NF"("UserID")
);
    
CREATE TABLE "Navigation_3NF"(
    "PathID" INTEGER NOT NULL, 
    "NavigationPath" VARCHAR(150),
    PRIMARY KEY("PathID" AUTOINCREMENT)
);


CREATE TABLE "PackageContents_3NF"(
    "ItemID" INTEGER NOT NULL,
    "ItemDescription" VARCHAR(150),
    "PackType" VARCHAR(150),
    "Warranty" VARCHAR(150),
    PRIMARY KEY("ItemID" AUTOINCREMENT)
);

CREATE TABLE "Colours_3NF"(
    "ColourID" INTEGER NOT NULL,
    "Colour" VARCHAR(150),
    PRIMARY KEY("ColourID" AUTOINCREMENT)
);

CREATE TABLE "Brands_3NF"(
    "BrandID" INTEGER NOT NULL,
    "Brand" VARCHAR(150),
    PRIMARY KEY ("BrandID" AUTOINCREMENT) 
);
    
CREATE TABLE "Locations_3NF"(
    "LocationID" INTEGER NOT NULL,
    "StockCountry" VARCHAR(150),
    PRIMARY KEY ("LocationID" AUTOINCREMENT)
);


Query started at 10:50:52 PM South Africa Standard Time; Query executed in 0.00 m

In [31]:
%%read_sql

--# Populate the tables to that they conform to the 3rd Normal Form
DELETE FROM "Products_3NF";
DELETE FROM "Transactions_3NF";
DELETE FROM "Carts_3NF";
DELETE FROM "Users_3NF";
DELETE FROM "Navigation_3NF";
DELETE FROM "PackageContents_3NF";
DELETE FROM "Colours_3NF";
DELETE FROM "Brands_3NF";
DELETE FROM "Locations_3NF";

INSERT INTO "Users_3NF" ("UserName")
SELECT DISTINCT UserName FROM Transactions_2NF;



Query started at 10:51:15 PM South Africa Standard Time; Query executed in 0.01 m

### <font color='turquoise'>Action: proceed with the insertion into `Carts_3NF`.   </font>

Take note that there are duplicate values coming from `Transactions_2NF` - retain the `distinct` combinations of `CartID`,`InvoiceDate`,`Total` only.

In [32]:
%%read_sql

INSERT INTO "Carts_3NF"("CartID","InvoiceDate","Total")
SELECT DISTINCT 
    CartID,
    InvoiceDate,
    Total
FROM Transactions_2NF;

Query started at 10:51:32 PM South Africa Standard Time; Query executed in 0.01 m

### <font color='turquoise'>Action: proceed with the insertion into the tables below. </font>

Similarly to the table above, mind the occurances of duplicate entries. These can be avoided by using `SELECT distinct` to avoid inserting redundant rows. 


In [33]:
%%read_sql

INSERT INTO "Locations_3NF" ("StockCountry")
SELECT DISTINCT 
    StockCountry 
FROM Products_2NF
WHERE Stockcountry <> '';

INSERT INTO "Brands_3NF" ("Brand")
SELECT DISTINCT 
    Brand 
FROM Products_2NF
WHERE Brand <> '';

INSERT INTO "Colours_3NF" ("Colour")
SELECT DISTINCT 
    Colour 
FROM Colours_2NF;

INSERT INTO "PackageContents_3NF" ("ItemDescription","PackType","Warranty")
SELECT DISTINCT 
    ItemDescription,
    PackType,
    Warranty 
FROM PackageContents_2NF;

INSERT INTO "Navigation_3NF" ("NavigationPath")
SELECT DISTINCT 
    NavigationPath 
FROM Navigation_2NF;



Query started at 10:51:52 PM South Africa Standard Time; Query executed in 0.01 m

### <font color='turquoise'>Action: proceed with the insertion into `Products_3NF`. </font>
Take note of the joins required for additional data contained in other tables. 

In [34]:
%%read_sql

INSERT INTO "Products_3NF" ("PathID","ItemID","ColourID","BrandID","LocationID",
                            "Barcode","ProductDescription","Price",
                            "Quantity","Weight_kg","Volume_litre",
                            "Length","Width","Height")
SELECT
    Navigation_3NF.PathID,
    PackageContents_3NF.ItemID,
    Colours_3NF.ColourID,
    Brands_3NF.BrandID,
    Locations_3NF.LocationID,
    Products_2NF.Barcode,
    ProductDescription,
    Price,
    Quantity,
    Weight_kg, 
    Volume_litre, 
    Length,
    Width,
    Height 
FROM 
    Products_2NF
LEFT JOIN Navigation_3NF ON Navigation_3NF.PathID=Products_2NF.PathID 
LEFT JOIN PackageContents_3NF ON PackageContents_3NF.ItemID=Products_2NF.ItemID
LEFT JOIN Colours_3NF ON Colours_3NF.ColourID=Products_2NF.ColourID
LEFT JOIN Brands_3NF ON Brands_3NF.Brand=Products_2NF.Brand
LEFT JOIN Locations_3NF ON  Locations_3NF.StockCountry=Products_2NF.StockCountry


Query started at 10:52:15 PM South Africa Standard Time; Query executed in 0.01 m

### <font color='turquoise'>Action: proceed with the insertion into `Transactions_3NF`.   </font>

Follow previous proceedure, and checks. 

In [35]:
%%read_sql

INSERT INTO "Transactions_3NF"("CartID||Barcode||UserName","CartID","Barcode","UserID")
SELECT DISTINCT 
     transactions_2nf.CartID||transactions_2nf.Barcode||Users_3NF.UserName,
     transactions_2nf.Barcode,
        transactions_2nf.CartID,
            Users_3NF.UserID
FROM
    Transactions_2NF
LEFT JOIN users_3NF ON Users_3NF.UserName=Transactions_2NF.UserName 
LEFT JOIN Carts_3NF ON Transactions_2NF.invoicedate = Carts_3nf.invoicedate
LEFT JOIN Products_3NF ON Transactions_2NF.Barcode = Products_3nf.Barcode

Query started at 10:53:13 PM South Africa Standard Time; Query executed in 0.01 m

### <font color='turquoise'>Action: Investigate the DB.   </font>

Note - the tables counts below are what you should arrive at. If you don't, reconsider how you have done the inserts above. Having the tables correctly formatted will allow to answer the questions more easily, and without errors.

| Table Name | Count |
| --- | --- |
| Brands_3NF | 232 |
| Carts_3NF |  190 |
| Colours_3NF | 17 |
| Locations_3NF | 2 |
| Navigation_3NF | 396 |
| PackageContents_3NF | 600 |
| Products_3NF | 1214 |
| Transactions_3NF | 275 |
| Users_3NF | 158 |

In [36]:
%%read_sql 

-- How many entries are here in each of the tables in the database now? 
select 'Products_3NF' as table_name, count(*) from Products_3NF
UNION
select 'Transactions_3NF' as table_name, count(*) from Transactions_3NF
UNION
select 'Users_3NF' as table_name, count(*) from Users_3NF
UNION
select 'Navigation_3NF' as table_name, count(*) from Navigation_3NF
UNION
select 'PackageContents_3NF' as table_name, count(*) from PackageContents_3NF
UNION
select 'Colours_3NF' as table_name, count(*) from Colours_3NF
UNION
select 'Brands_3NF' as table_name, count(*) from Brands_3NF
UNION
select 'Locations_3NF' as table_name, count(*) from Locations_3NF
UNION
select 'Carts_3NF' as table_name, count(*) from Carts_3NF


Query started at 10:53:25 PM South Africa Standard Time; Query executed in 0.00 m

table_name  count(*)
0           Brands_3NF       232
1            Carts_3NF       190
2          Colours_3NF        17
3        Locations_3NF         2
4       Navigation_3NF       396
5  PackageContents_3NF       600
6         Products_3NF      1214
7     Transactions_3NF       275
8            Users_3NF       158

<a id='MCQ_questions'></a>
## 9. MCQ Questions

Having completed the normalisation of the database, you may use the following cells to help you answer the MCQs for the predict. 

Q1) How many unique products does the company have?

In [37]:
%%read_sql
-- Write your query here:
SELECT COUNT(DISTINCT Barcode) AS 'Products'
FROM Products_3NF;

Query started at 10:54:11 PM South Africa Standard Time; Query executed in 0.00 m

Products
0       608

Q2) How many users bought from Bhejane in April 2020?

In [38]:
%%read_sql
-- Write your query here:
select Count(DISTINCT username) AS "Users"
from transactions
WHERE Invoicedate BETWEEN '2020-04-01' AND '2020-05-01'

Query started at 10:54:26 PM South Africa Standard Time; Query executed in 0.00 m

Users
0     31

Q3) How many users bought 3 or more items that cost more than R1000?

In [58]:
%%read_sql
-- Write your query he
SELECT c.CartID,c.InvoiceDate,Transactions.*
from Carts_3NF as c
INNER JOIN 
Transactions 
ON c.CartID = Transactions.CartID
WHERE Transactions.Total > 1000  
ORDER BY Transactions.UserName DESC



Query started at 11:47:12 PM South Africa Standard Time; Query executed in 0.00 m

CartID         InvoiceDate  CartID        Barcode    Total       UserName  \
0      179  2020-04-04 0:00:00     179   617566827837   3464.0         Zanele   
1      179  2020-04-04 0:00:00     179    HTBH54I571A   3464.0         Zanele   
2      179  2020-04-04 0:00:00     179  MPTAL72013338   3464.0         Zanele   
3       23  2020-03-28 0:00:00      23  5397184200193   2299.0         Todani   
4      126  2020-08-12 0:00:00     126    HTB5FI3BC8C   3765.0        Mzoxolo   
5      113  2020-08-11 0:00:00     113  6003000640348   2535.0        Melandi   
6      113  2020-08-11 0:00:00     113  8001841126586   2535.0        Melandi   
7      113  2020-08-11 0:00:00     113    MPTALP14631   2535.0        Melandi   
8      172  2020-07-24 0:00:00     172   884392580773   1114.0            Mel   
9      172  2020-07-24 0:00:00     172  6009900463131   1114.0            Mel   
10     172  2020-07-24 0:00:00     172  MPTAL57927480   1114.0            Mel   
11     172  2020-07-24 0:00:00     172  MPTAL72842868   1114.0            Mel   
12     115  2020-06-13 0:00:00     115   834266002368   1716.0       Margaret   
13     115  2020-06-13 0:00:00     115  6009684040641   1716.0       Margaret   
14     115  2020-06-13 0:00:00     115  8886362979725   1716.0       Margaret   
15     115  2020-06-13 0:00:00     115  MPTAL00031884   1716.0       Margaret   
16     156  2020-06-07 0:00:00     156   718037858494   1351.0         Mandla   
17     123  2020-04-11 0:00:00     123    HTB5ABG31E3  10999.0         Mandla   
18     156  2020-06-07 0:00:00     156  MPTAL00348002   1351.0         Mandla   
19     156  2020-06-07 0:00:00     156  MPTAL00573464   1351.0         Mandla   
20     114  2020-05-31 0:00:00     114   884392580773   1137.0          Karyn   
21     114  2020-05-31 0:00:00     114  6002322014646   1137.0          Karyn   
22     114  2020-05-31 0:00:00     114  MPTAL57567320   1137.0          Karyn   
23     136  2020-08-04 0:00:00     136   619659097318   3301.0         Junaid   
24     136  2020-08-04 0:00:00     136  8809313492623   3301.0         Junaid   
25     136  2020-08-04 0:00:00     136    HTBCGGHCC14   3301.0         Junaid   
26     136  2020-08-04 0:00:00     136  MPTAL00351754   3301.0         Junaid   
27     101  2020-06-09 0:00:00     101  4549292118766   1334.0          Hanno   
28     101  2020-06-09 0:00:00     101  6001878010171   1334.0          Hanno   
29     101  2020-06-09 0:00:00     101  MPTAL00215519   1334.0          Hanno   
30     160  2020-04-17 0:00:00     160  4015400541813   1134.0          Daffy   
31     160  2020-04-17 0:00:00     160  6001068495306   1134.0          Daffy   
32     160  2020-04-17 0:00:00     160  MPTAL00404508   1134.0          Daffy   
33     160  2020-04-17 0:00:00     160  MPTAL72134564   1134.0          Daffy   
34     102  2020-07-02 0:00:00     102      300507946   1523.0         DIMPHO   
35     102  2020-07-02 0:00:00     102  6002322009727   1523.0         DIMPHO   
36     102  2020-07-02 0:00:00     102  8888021200980   1523.0         DIMPHO   
37     102  2020-07-02 0:00:00     102  MPTAL00215561   1523.0         DIMPHO   
38       7  2020-03-31 0:00:00       7   741311301174   1999.0          Cindy   
39     124  2020-05-30 0:00:00     124    HTB5ABG31E3  10999.0       Brigette   
40      57  2020-06-17 0:00:00      57  6005519114211   1999.0      Anonymous   
41     125  2020-03-27 0:00:00     125    HTB5FI3BC8C   3765.0  Andreas Peter   
42     139  2020-06-27 0:00:00     139  6009706869533   1177.0          Adele   
43     139  2020-06-27 0:00:00     139    HTBG3HG7A86   1177.0          Adele   
44     139  2020-06-27 0:00:00     139  MPTAL00451356   1177.0          Adele   

           InvoiceDate  
0   2020-04-04 0:00:00  
1   2020-04-04 0:00:00  
2   2020-04-04 0:00:00  
3   2020-03-28 0:00:00  
4   2020-08-12 0:00:00  
5   2020-08-11 0:00:00  
6   2020-08-11 0:00:00  
7   2020-08-11 0:00:00  
8   2020-07-24 0:00:00  
9   2020-07-24 0:00:00  
10

Q4) Which user made the largest purchase on a single transaction?

In [47]:
%%read_sql
-- Write your query here:
SELECT Users_3NF.UserName,Total
FROM Carts_3NF
LEFT JOIN Transactions_3NF
USING(CartID)
LEFT JOIN Users_3NF
USING(UserID)
ORDER BY Total DESC

Query started at 11:00:15 PM South Africa Standard Time; Query executed in 0.00 m

UserName    Total
0       None  10999.0
1       None  10999.0
2       None   3765.0
3       None   3765.0
4       None   3464.0
..       ...      ...
185     None     23.0
186     None     13.0
187     None     12.0
188     None     10.0
189     None      1.0

[190 rows x 2 columns]

Q5) How many components does the product: "5m Colour Changing RGB LED Strip Light" (MPTAL57588104) come with?

In [48]:
%%read_sql
-- Write your query here:
SELECT COUNT(DISTINCT Itemdescription) AS "Components"
FROM PackageContents_3NF
LEFT JOIN Products_3NF
USING(ItemID) 
WHERE productdescription= "5m Colour Changing RGB LED Strip Light" 

Query started at 11:12:07 PM South Africa Standard Time; Query executed in 0.00 m

Components
0           4

Q6) How many brands are available at Bhejane?

In [ ]:
%%read_sql
-- Write your query here:
SELECT COUNT(DISTINCT Brand) AS "Bhejane Brands"
FROM Brands_3NF%%read_sql
-- Write your query here:


7) What is the price of the "Verimark - Floorwiz 2in1 Mop"?

In [49]:
%%read_sql
-- Write your query here:
SELECT productdescription,price
FROM products_3NF
WHERE Productdescription='Verimark - Floorwiz 2in1 Mop'
LIMIT 2;

Query started at 11:12:38 PM South Africa Standard Time; Query executed in 0.00 m

ProductDescription  Price
0  Verimark - Floorwiz 2in1 Mop  179.0
1  Verimark - Floorwiz 2in1 Mop  179.0

Q8) Calculate the package volume of the "Russell Hobbs - Slow Cooker" using the given dimensions

In [ ]:
%%read_sql
-- Write your query here:
WITH CTE AS(
SELECT Length*Width*Height as Volume,productdescription
FROM Products_3NF)
SELECT *
FROM CTE
WHERE productdescription='Russell Hobbs - Slow Cooker'
LIMIT 2;

Q9) Which user made the most transactions in the Year 2020?

In [51]:
%%read_sql
SELECT Barcode,COUNT(UserName) as Users2020, UserName,InvoiceDate
FROM Transactions
WHERE InvoiceDate BETWEEN "2020-01-01 0:00:00" AND "2020-12-31 0:00:00"
GROUP BY UserName
ORDER BY Users2020 DESC

Query started at 11:18:55 PM South Africa Standard Time; Query executed in 0.00 m

Barcode  Users2020            UserName         InvoiceDate
0    4015400541813          9            Cornelis  2020-05-19 0:00:00
1     790069435478          8                Erlo  2020-04-10 0:00:00
2     617566827837          6              Zanele  2020-04-04 0:00:00
3    6003977000206          6             Heather  2020-04-22 0:00:00
4     718037858494          5              Mandla  2020-06-07 0:00:00
..             ...        ...                 ...                 ...
155  6007226069631          1  Angelica Fernandez  2020-05-29 0:00:00
156  6001106124069          1             Andiswa  2020-08-08 0:00:00
157  6002322013823          1   Andisa Moleboheng  2020-06-11 0:00:00
158  5010103800303          1              Alicia  2020-05-15 0:00:00
159  8001090660008          1             Adeebah  2020-07-12 0:00:00

[160 rows x 4 columns]

Q10) What is the total number of users that shop at Bhejane?

In [52]:
%%read_sql
-- Write your query here:
SELECT COUNT(DISTINCT UserID)
FROM Users_3NF

Query started at 11:19:14 PM South Africa Standard Time; Query executed in 0.00 m

COUNT(DISTINCT UserID)
0                     158

Q11) What is the record count for the Colours_3NF Table?

In [53]:
%%read_sql
-- Write your query here:
SELECT COUNT(*) AS "Records"
FROM Colours_3NF

Query started at 11:19:30 PM South Africa Standard Time; Query executed in 0.00 m

Records
0       17

Q12) What would the total price be if I had the following items in my cart?
* MPTAL57588104
* 5000394203921
* 6932391917652

In [54]:
%%read_sql
-- Write your query here:
SELECT SUM(DISTINCT Price) AS "Total Price"
FROM Products_3NF
WHERE Barcode in('MPTAL57588104','5000394203921','6932391917652')
limit 10

Query started at 11:19:51 PM South Africa Standard Time; Query executed in 0.00 m

Total Price
0        734.0

Q13) What is the barcode of the most sold product?

In [ ]:
%%read_sql
SELECT count( Barcode) as BC, Barcode from Transactions
group by Barcode
order by BC desc

Q14) What are the products of Cornelis’ cart on the 2020-06-28 0:00:00?

In [56]:
%%read_sql
-- Write your query here:
SELECT  p.Barcode,p.ProductDescription,t.InvoiceDate,t.CartID,t.Barcode,t.UserName
from Products_3NF as p
INNER JOIN Transactions t
ON p.Barcode = t.Barcode
WHERE t.InvoiceDate = "2020-06-28 0:00:00" AND t.UserName = "Cornelis"

Query started at 11:29:58 PM South Africa Standard Time; Query executed in 0.00 m

Barcode                                 ProductDescription  \
0    6001106124069   Dettol Body Wash - Shower Gel - Original - 600ml   
1  6003753002035AP                       Vitality Aloe Vera Gel 100ml   
2  6003753002035AP                       Vitality Aloe Vera Gel 100ml   
3  6003753002035AP                       Vitality Aloe Vera Gel 100ml   
4    MPTAL00555756  Sparq Active Resistance Loop Band - Set of 5 w...   
5    MPTAL00555756  Sparq Active Resistance Loop Band - Set of 5 w...   
6    MPTAL72193414                             Protective Face Shield   
7    MPTAL72193414                             Protective Face Shield   

          InvoiceDate  CartID          Barcode  UserName  
0  2020-06-28 0:00:00     104    6001106124069  Cornelis  
1  2020-06-28 0:00:00     104  6003753002035AP  Cornelis  
2  2020-06-28 0:00:00     104  6003753002035AP  Cornelis  
3  2020-06-28 0:00:00     104  6003753002035AP  Cornelis  
4  2020-06-28 0:00:00     104    MPTAL00555756  Cornelis  
5  2020-06-28 0:00:00     104    MPTAL00555756  Cornelis  
6  2020-06-28 0:00:00     104    MPTAL72193414  Cornelis  
7  2020-06-28 0:00:00     104    MPTAL72193414  Cornelis

Q15) Which users bought locally produced, black-coloured products on odd-numbered months of the year, and what was the total cost (rounded to the nearest integer) of the carts containing these products?

In [57]:
%%read_sql
-- Write your query here:
SELECT Products_1NF.Stockcountry,Transactions_1NF.Username,Products_1NF.Colour,Transactions_1NF.Invoicedate,Transactions_1NF.Total
FROM Products_1NF
LEFT JOIN Transactions_1NF
USING(Barcode)
WHERE Stockcountry='South Africa' and Colour='Black' and Invoicedate like '%05%'

Query started at 11:30:42 PM South Africa Standard Time; Query executed in 0.00 m

StockCountry UserName Colour         InvoiceDate  Total
0  South Africa    Carel  Black  2020-05-04 0:00:00  799.0
1  South Africa    Carel  Black  2020-05-04 0:00:00  799.0
2  South Africa    Carel  Black  2020-05-04 0:00:00  799.0
3  South Africa    Carel  Black  2020-05-04 0:00:00  799.0
4  South Africa    Lacin  Black  2020-05-18 0:00:00  798.0
5  South Africa    Lacin  Black  2020-05-18 0:00:00  798.0